.cbmc-batch/jobs/aws_cryptosdk_serialize_frame/aws_cryptosdk_serialize_frame_harness.c

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/byte_buf.h>
+#include <aws/cryptosdk/private/framefmt.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void harness() {
+    /* data structure */
+    struct aws_cryptosdk_frame frame;
+    size_t ciphertext_size;
+    size_t plaintext_size;
+    struct aws_byte_buf ciphertext_buf;
+    struct aws_cryptosdk_alg_properties alg_props;
+
+    /* Assumptions about the function input */
+    ensure_byte_buf_has_allocated_buffer_member(&ciphertext_buf);
+    __CPROVER_assume(aws_byte_buf_is_valid(&ciphertext_buf));
+
+    ensure_alg_properties_has_allocated_names(&alg_props);
+    __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(&alg_props));
+
+    /* Save the old state of the ciphertext buffer */
+    uint8_t *old_ciphertext_buffer   = ciphertext_buf.buffer;
+    size_t old_ciphertext_buffer_len = ciphertext_buf.len;
+
+    int rval = aws_cryptosdk_serialize_frame(&frame, &ciphertext_size, plaintext_size, &ciphertext_buf, &alg_props);
+    if (rval == AWS_OP_SUCCESS) {
+        assert(aws_cryptosdk_frame_is_valid(&frame));
+        assert(ciphertext_buf.buffer == old_ciphertext_buffer);
+        assert(ciphertext_buf.len == old_ciphertext_buffer_len + ciphertext_size);
```

Comment:
As you found, `aws_cryptosdk_serialize_frame` modifies the ciphertext size to track the total ciphertext created.
If you want to make this assertion, rather than changing the behavior of `aws_cryptosdk_serialize_frame` to set the ciphertext size to zero (and thus making this line a no-op),
you should be making a more correct assertion about what is actually supposed to be happening here.

One complication here is that `ciphertext_buf.len` is going to increase *not only* by the length of the encrypted ciphertext (equal to the plaintext size) but also by the lengths of the frame metadata.

Is there a helper function somewhere that will determine the size of a frame given the frame length, frame type, and algorithm suite?

Reply:
I didn't find any helper function that returns the size. We could probably derive one though based on the calls to `field_sized` and `field_be32/64` in `serde_framed`, `serde_last_frame`, and `serde_nonframed` (such as [this call](https://github.com/aws/aws-encryption-sdk-c/pull/378/files#diff-72d57ea9d90face9d6fb936b9bbbed84R169)). 

In general, before the modification that I did to the code, the returned `ciphertext_size` had the value `plaintext_size + ciphertext_buf.len - old_ciphertext_buf.len`.

Reply:
In PR #379 for the deserialize function I have manually computed the size of the frame for each case. I can extract them into helper functions if those don't already exist.

Reply:
@ttjsu-aws Is this assertion reasonable? If not, then what does the returned `ciphertext_size` represent?

Reply:
The ciphertext_size represents the amount of ciphertext to write to the frame and this is the value returned by the `aws_cryptosdk_serialize_frame` function.  `ciphertext_buf` is tracking the cursor for the frames initialized so far.  Given that, this assertion is not a reasonable one.

Reply:
@ttjsu-aws  Thanks for the response! However, there is something here that still confuses me. The returned ciphertext size is increased every time some space is reserved in the ciphertext buffer (see the macro definition `field_sized(state, bufptr, size)`). What is a reasonable assertion for the returned `ciphertext_size` value if the call to `serialize` succeeded?

Reply:
The returned ciphertext_size is increased by the size of the IV/authtag (if the frame has one). The IV/authtag size is not included in the ciphertext_size earlier.  Refer [framefmt.c#L168 ](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L168) and [framefmt.c#L170 ](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L170 for framed data and [framefmt.c#L203](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L203) and [framefmt.c#L206](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L206)

Reply:
@ttjsu-aws  Isn't the `plaintext_size` added again in [https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L205](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L205) for nonframed and in [https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L169](https://github.com/aws/aws-encryption-sdk-c/blob/master/source/framefmt.c#L169) for framed? In total ciphertext_size will contain the plaintext_size twice, once in the initialization and once after calling `field_sized`.

Reply:
If that is the expected behaviour, is the correct assertion for the returned ciphertext_size `ciphertext_size == new_ciphertext_buf_len - old_ciphertext_buf_len + plaintext_size`?

Reply:
I seem to have misunderstood the problem with the field_sized and field_header macros, I trust that Bryan and Dave's comments have resolved this  questions.

Reply:
Thanks, they have